### ISSUE TITLE
'EXAODBC' driver must be defined although using DSN with driver specified

### ISSUE THREAD

**Issue by seanv507** [exasol/sqlalchemy-exasol#68]
Hi
if I use a DSN (with driver specified) then I still need EXAODBC driver defined (in odbcinst.ini)

engine = sqlalchemy.create_engine('exa+turbodbc://@exasolution-uo2214lv2_64')
then query results in 

DatabaseError: (turbodbc.exceptions.DatabaseError) ODBC error
state: 01000
native error code: 0
message: [unixODBC][Driver Manager]Can't open lib 'EXAODBC' : file not found

As soon as I define an odbcinst.ini with EXAODBC driver, problem goes away.

**Comment by jank** [comment]
can you share (anonymized) the DSN configuration and odbcins.ini configuration you are using?

**Comment by seanv507** [comment]
here you go

[EDIT I had attached them, but github screened them out because they have the 'ini' file extension]

My point is that the DSN details already contains the driver information, so why do I need a separate odbcinst.ini file referring to EXAODBC?  I don't need the odbcinst file if eg i use isql I can just use odbc.ini.

and so I can connect by adding a ODBCINST.ini file but it seems 'wrong' to have to add it instead of using the DSN details

the odbc.ini file is generated by running the exasolution odbc config_odbc script....


odbc.ini
[ODBC Data Sources]
exasolution-uo2214lv2_64 = unixODBC 2.2.14 or later, libversion 2 (64bit)


[exasolution-uo2214lv2_64]
DRIVER = XXXX/EXASolution_ODBC-5.0.17/lib/linux/x86_64/libexaodbc-uo2214lv2.so
EXAHOST = XXXX
EXAUID = XXXX
EXAPWD = XXXX
EXASCHEMA = XXXX
CONNECTIONLCALL = en_US.utf-8

[exasolution-uo2214lv2_64-debug]
DRIVER = XXXX/EXASolution_ODBC-5.0.17/lib/linux/x86_64/libexaodbc-uo2214lv2.so
EXAHOST = XXXX
EXAUID = XXXX
EXAPWD = XXXX
EXASCHEMA = XXXX
CONNECTIONLCALL = en_US.utf-8
EXALOGFILE = XXXX/EXASolution_ODBC-5.0.17/exaodbc.log
LOGMODE = verbose


odbcinst.ini

[EXAODBC]
DRIVER = XXXX/EXASolution_ODBC-5.0.17/lib/linux/x86_64/libexaodbc-uo2214lv2.so
Threading=2

**Comment by seanv507** [comment]
I may have got the wrong end of the stick [i have no knowledge of sqlalchemy:)] but a quick debug showed this:

anaconda3/lib/python3.6/site-packages/sqlalchemy/engine/default.py(393)connect()
    391 
    392     def connect(self, *cargs, **cparams):
--> 393         return self.dbapi.connect(*cargs, **cparams)
    394 
    395     def create_connect_args(self, url):

ipdb> cargs
('exasolution-uo2214lv2_64',)
ipdb> cparams
{'driver': 'EXAODBC', 'inttypesinresultsifpossible': 'y', 'turbodbc_options': <turbodbc_intern.Options object at 0x7f0a6ce3be68>}

**Comment by jank** [comment]
What I can read / deduce from the traces is:

* you are using the DSN 'exasolution-uo2214lv2_64' - can you share details of the respective DSN config?
* the turbodbc dialect (sqlalchemy term) uses the string 'EXAODBC' as default to look for a driver configuration in the odbcinst.ini, this often confuses people. EXAODBC is not the driver name, it is just an identifier used by ODBC to look for a configuration entry with that name in the odbcinst.ini

a sample configuration that wires the concrete driver could look like this:

```
[EXAODBC]
DRIVER=some/directory/driver/libexaodbc-uo2214.so
```

So I assume that everything works as expected. All the config files, defaults, and indirections are just quite confusing. We actually considered making this more explicit by implementing https://github.com/blue-yonder/sqlalchemy_exasol/issues/66

Please let me know if this helps you achieving your connectivity.

**Comment by seanv507** [comment]
Hi BY-jk 
I have responded to your comments in my message above (where I had tried to attach the ini files)

**Comment by seanv507** [comment]
My limited undestanding of ODBC.ini files and odbcinst.ini files comes from 
http://turbodbc.readthedocs.io/en/latest/pages/odbc/driver_manager_config.html, which as I say seems to suggest that the odbcinst.ini file is only necessary if you use driver connection string (and even there one is not limited to calling it EXAODBC.) rather than dsn connection.

**Comment by jank** [comment]
As I see it, the inclusion of the DEFAULT driver

https://github.com/blue-yonder/sqlalchemy_exasol/blob/99b7e0d0b8dae81928b449b14fcb338eb2eef34c/sqlalchemy_exasol/turbodbc.py#L97

forces unixodbc to lookup the driver in the odbcinst.ini instead of using the DRIVER parameter provided with the DSN. 

This all relates back to #66. Once we would remove the default behavior, this behavior would go away. This is relatively little work on this dialect but more involved to all users as this would be a breaking change.

Still, given that you were the next to run into this trap we'll have to revisit this behavior.

Until #66 is closed your choices are:
* pass the driver in the user options
* define the drive in odbcinst.ini under section EXAODBC

In the current state I do not see a proper way of reading out the DRIVER parameter of the DSN configuration.

**Comment by seanv507** [comment]
Thanks Jan
I just wanted to raise this issue ( I have got a workaround), so please
close/leave open as you see fit..

**Comment by jank** [comment]
Closing as workaround exists - watch / comment on #66

**Comment by jank** [comment]
Branch sqla1_2 includes the fix for #66. This branch will be merged to master and sqlalchemy_exasol 2.0.0 released after a stable SQLA 1.2.0 version is available.
